dali/test/dali_operator_test.h

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_TEST_DALI_OPERATOR_TEST_H_
+#define DALI_TEST_DALI_OPERATOR_TEST_H_
+
+#include <gtest/gtest.h>
+#include <vector>
+#include "tensor_adapter.h"
+#include "operators_graph.h"
+
+namespace dali {
+
+namespace testing {
+
+template<typename InputType, typename OutputType>
+class DaliOperatorTest : public ::testing::Test {
+  static_assert(std::is_fundamental<InputType>::value, "DaliOperatorTest expects fundamental type as InputType");
+  static_assert(std::is_fundamental<OutputType>::value, "DaliOperatorTest expects fundamental type as OutputType");
+
+public:
+  using Arguments =  std::map<std::string, int>;  // TODO some generalization (boost::any? tagged union?)
+
+  void RunTest(Arguments operator_arguments, std::vector<TensorAdapter<OutputType>> anticipated_outputs) noexcept {}
+
+private:
+  virtual std::vector<TensorAdapter<InputType>> SetInputs() const noexcept = 0;
```

Comment:
1. `SetInputs() const` is a bad name. What does this function do? Shouldn't it be `GetInputs` or `GenerateInputs`?
2. Private pure virtual function makes the class useless.
3. Why doesn't it work the other way round? I.e. `SetInputs(const std::vector<TensorAdapter<InputType>> &)`?

Reply:
ad 1. It was made for reasons provided below (for the templated method pattern). If you feel, that it's inaccurate, I can change it to `GenerateInputs`
ad 2, ad 3. It's [template method](https://en.wikipedia.org/wiki/Template_method_pattern). Not to be confused with _method template_ as in
`template<typename T> void method(T t);` ;-)
I've pushed a little code into `SetUp()` so that it'd be clearer, what's the point of private virtual methods here

Reply:
`SetUp` is a template method. `GenerateInputs` is part of its implementation. And if it's meant for overriding, it's an API - and having API in the form of a private method smells bad. A method (or field, for that matter) should be private, if its invocation (or altering its value) can destroy object's consistency. If it doesn't, it should be public.

BTW I don't like this template methods (the final ones). They impose some fixed lifecycle of the test and  this is likely to explode in our faces sooner or later.